### PR TITLE
Issue 3364 - fix for a void-cast queue pop_front or pop_back when it's the first SV statement in a block of code.

### PR DESCRIPTION
**Pull request by drewranck**
#3364 
This is a fix for SystemVerilog like:
if (foo.size()) void'(foo.pop_front());

Which previously would not compile due to generated .cpp missing a ;

I'm not crazy about my method names added in V3Ast.h, and am open to suggestions. If you think this is the wrong fix, then I'll remove the PR and think about it some more.

### CODE REVIEW

**Comment by drewranck** [comment]
As someone w/ only gcc, I didn't realize I needed the 'override' keyword in virtual functions... will try again :(

File: src/V3Ast.h [hunk 1]
```diff
@@ -1509,6 +1509,16 @@ class AstNode VL_NOT_FINAL {
     AstNode* firstAbovep() const {  // Returns nullptr when second or later in list
         return ((backp() && backp()->nextp() != this) ? backp() : nullptr);
     }
+    // isFirstInOneOfMyListOfStatements(n) -- is implement for base classes:
+    // AstNodeBlock, AstCaseItem, AstNodeIf, AstNodeFTask.
+    virtual bool isFirstInOneOfMyListOfStatements(AstNode* n) const { return false; }
```
Comment (wsnyder) [inline src/V3Ast.h:1514]:
```suggestion
    virtual bool isFirstInMyListsOfStatements(AstNode* nodep) const { return false; }
```

File: src/V3Ast.h [hunk 2]
```diff
@@ -2344,6 +2354,9 @@ class AstNodeBlock VL_NOT_FINAL : public AstNode {
     AstNode* stmtsp() const { return op1p(); }  // op1 = List of statements
     void addStmtsp(AstNode* nodep) { addNOp1p(nodep); }
     bool unnamed() const { return m_unnamed; }
+    virtual bool isFirstInOneOfMyListOfStatements(AstNode* n) const override {
+        return (n != nullptr && n == stmtsp());
+    }
```
Comment (wsnyder) [inline src/V3Ast.h:2359]:
```suggestion
    virtual bool isFirstInOneOfMyListOfStatements(AstNode* nodep) const override {
        return (nodep && nodep == stmtsp());
    }
```

And rename method name, likewise elsewhere

**Review by wsnyder — COMMENTED** [review]
I think this is a great approach.

I'd suggest add to V3Broken a check that if there's a NodeStmt it should always have a parent (backp()) that has isStandloneBodyStmt() true.  I suspect this will find some other missing nodes which need the new virtual method.  With those fixed I think we'll then have a great fix that is fully robust.

Thanks for your work on this - appreciate it's grown to more effort than I originally thought it would be!

**Comment by drewranck** [comment]
Sorry, this totally fell off my radar for months, and I'll take a look to rework this PR.

**Comment by drewranck** [comment]
> I think this is a great approach.
> 
> I'd suggest add to V3Broken a check that if there's a NodeStmt it should always have a parent (backp()) that has isStandloneBodyStmt() true. I suspect this will find some other missing nodes which need the new virtual method. With those fixed I think we'll then have a great fix that is fully robust.
> 
> Thanks for your work on this - appreciate it's grown to more effort than I originally thought it would be!

I'm not too familiar with how V3Broken works. I'm also not sure that every AstNodeStmt should have a parent with isStandaloneBodyStmt() == true. Maybe this is a confusing naming choice on my end because isStandaloneBodyStmt() is really telling us if we need a generated .cpp to add a semicolon to the statement (meaning, for example, it's not a conditional expression statement, it's a "body" statement). 

I agree that every AstNodeStmt should have a parent (backp()) though. I'll look into this.

Also, it appears my delay on working on the issue has left my branch really far behind. I rebased onto a new branch from verilator:master (https://github.com/drewranck/verilator/tree/issue-3364-try3) , which means I think I have to start the PR over again.

**Comment by wsnyder** [comment]
Merged with #3364, this one is stale.

**Pull request closed without merging**